twpca/crossval.py

```diff
@@ -0,0 +1,183 @@
+"""
+TWPCA cross-validation and hyperparameter search routines
+"""
+import numpy as np
+import tensorflow as tf
+import itertools
+from tqdm import tqdm
+from .model import TWPCA
+from .regularizers import curvature
+from .utils import stable_rank
+
+__all__ = ['cross_validate', 'hyperparam_search', 'err_per_component']
+
+def k_fold(N, K):
+    """K-fold cross validation iterator
+
+    Args:
+        N: int, number of datapoitns
+        K: int, number of folds
+    """
+    idx = np.arange(N)
+    for k in range(K):
+        train_idx = idx[idx % K != k]
+        test_idx = idx[idx % K == k]
+        yield train_idx, test_idx
+
+def leave_n_out(N, K):
+    """Leave K out cross validation iterator
+
+    Args:
+        N: int, number of datapoitns
+        K: int, number to leave out in training set
+    """
+    idx = np.arange(N)
```

Comment:
I was back and forth about which would be better. If we're doing this over multiple hyperparams we might want the same random permutation... I think I'll fix by passing a seed.